rosbag2_transport/src/rosbag2_transport/player.cpp

```diff
@@ -254,6 +253,49 @@ bool Player::set_rate(double rate)
   return clock_->set_rate(rate);
 }
 
+rosbag2_storage::SerializedBagMessageSharedPtr * Player::peek_next_message_from_queue()
+{
+  rosbag2_storage::SerializedBagMessageSharedPtr * message_ptr = message_queue_.peek();
+  if (message_ptr == nullptr && !is_storage_completely_loaded() && rclcpp::ok()) {
+    RCLCPP_WARN(
+      this->get_logger(),
+      "Message queue starved. Messages will be delayed. Consider "
+      "increasing the --read-ahead-queue-size option.");
+    while (message_ptr == nullptr && !is_storage_completely_loaded() && rclcpp::ok()) {
+      std::this_thread::sleep_for(std::chrono::microseconds(100));
+      message_ptr = message_queue_.peek();
+    }
+  }
+  return message_ptr;
+}
+
+bool Player::play_next()
+{
+  if (!clock_->is_paused()) {
```

Comment:
`playing_messages_from_queue_`  was introduced purely to solve non-determinism in tests. 
From first sight `playing_messages_from_queue_` this is exactly from `play_next()` depend upon. 
But it will be difficult to describe and meet this condition from upper API level. i.e. Player::play() method call. 
We might have already call `Player::play()` method but haven't yet reached the place where `playing_messages_from_queue_` triggered to true. In this case `play_next()` will return `false` which is sort of unexpected   and caused by race conditions.
I proposed to guard it with another flag `is_in_play_` to avoid such race conditions. You might argue that it will over-complicate the logic insight player class. I agree with it and this is why I put constrained about prior call for `play()` in description for `play_next()` method from the first hand.

Reply:
Sure, that makes sense. If we can't make `playing_messages_from_queue_` do this, I would prefer to have the extra condition, rather than undefined behavior